### PR TITLE
typescript: a new parser

### PR DESCRIPTION
**Pull request by ksamborski**
Hello!

I've implemented basic support for typescript. It includes following tag kinds:

- a: type aliases
- c: classes
- e: enums
- i: interfaces
- f: functions
- g: generators
- m: class methods
- n: namespaces
- p: class properties
- v: variables

Tags are generated with additional access and scope fields. Currently signatures are not yet gathered.
I tested this parser on different codebases, such as [NestJS framework](https://github.com/nestjs/nest) and some proprietary ones at work. Whatever I found I added to the unit tests. When I tested it manually I did it just by looking in Tagbar in vim. It seems to work well.

Anyway please review my work. This is my first PR here and it was quite a challenge.

### CODE REVIEW

**Comment by masatake** [comment]
A hero comes!
30 people have waited for this in #1566:-)

Your .ts test cases are failed on ci/cd environments.
It seems that your parser calls too much ungetc.

```
ctags: main/read.c:786: ungetcToInputFile: Assertion `File.ungetchIdx < len' failed.
	ctags: main/read.c:786: parsing ./Units/parser-typescript.r/ts-namespaces.d/input.ts:3 as TypeScript
```

Could you try --enable-debugging option of ./configure?
This option activates Assert macros.

```
$ make clean
$ ./configure --enable-debugging
$ make
$ make units LANGUAGES=TypeScript
```

expected.tags files become easier to read with putting args.ctags with following contents to each test case directories:
```
--sort=no
```

**Comment by ksamborski** [comment]
Oh, I didn't know about the `--enable-debugging` option. I'll will fix the errors, thanks!

**Comment by masatake** [comment]
BTW, If you have valgrind memory debugger, you can run "make units" under valgrind with following command line:
```
$ make units LANGUAGES=TypeScript VG=1
```

**Comment by ksamborski** [comment]
Yes, I know. I run valgrind very often.

**Comment by ksamborski** [comment]
I was tying to implement simple backtracking mechanism but I see now that `ungetchBuf` is only 8 elements long. Can we just change it to something like 80 or should I implement this differently? BTW. 14 is enough for tests to pass.

**Comment by masatake** [comment]
I guess 80 can be not-enough.
You are implementing backing track mechanism on byte stream. So you need non-limit ungetc.
Isn't it possible to implement backing track mechanism on token stream?
So you can control the length (depth?) as you want.

See https://github.com/universal-ctags/ctags/blob/master/main/tokeninfo.h

I implemented generic token structure.
I implemented 
```
void tokenUnread (tokenInfo *token);
```.

I used the function in https://github.com/universal-ctags/ctags/blob/master/parsers/ldscript.c .

**Comment by masatake** [comment]
...using CPreProcessor may not be a good idea.

**Comment by ksamborski** [comment]
It may not be a good idea to use it directly but those functions actually do exactly what I need. Maybe could I extract those functions into core or something?

**Comment by masatake** [comment]
...I wonder getInputLineNumber() and  getInputFilePosition() are reliable or not.

**Comment by masatake** [comment]
If you use CPreProcessor  you may have to deal with values which you cannot store to char.
```
enum eCharacters {
	/* white space characters */
	SPACE         = ' ',
	NEWLINE       = '\n',
	CRETURN       = '\r',
	FORMFEED      = '\f',
	TAB           = '\t',
	VTAB          = '\v',

	/* some hard to read characters */
	DOUBLE_QUOTE  = '"',
	SINGLE_QUOTE  = '\'',
	BACKSLASH     = '\\',

	/* symbolic representations, above 0xFF not to conflict with any byte */
	STRING_SYMBOL = ('S' + 0xff),
	CHAR_SYMBOL   = ('C' + 0xff)
};
```
See STRING_SYMBOL and CHAR_SYMBOL.
Anyway,  storing the value returned from getcFromInputFile() to char variable is not good.
int is needed.

I will put review comments.

File: parsers/typescript.c [hunk 1]
```diff
@@ -0,0 +1,2001 @@
...
+BLOCK_PARSER_DEF (Curlies, '{', '}', TOKEN_CURLIES)
+
+static bool tryParser(Parser parser, ParserStateInit stInit, ParserStateFree stFree, tokenInfo *const token)
+{
+	void *currentState = NULL;
+	parserResult result;
+	result.status = PARSER_NEEDS_MORE_INPUT;
+	result.unusedChars = 0;
+	charArray *usedC = charArrayNew ();
+	char c;
```
Comment (masatake) [inline parsers/typescript.c]: int is needed for c.

Functions for reading input returns int, not char. There is a reason.
These function may return 0  ~ 255 or EOF.
If the platform that we use uses singned char as "char", we cannot receive value 255. 

```
[yamato@slave]/tmp% cat foo.c
cat foo.c
#include <stdio.h>

int f()
{
  return 255;
}

int e()
{
  return EOF;
}

int main(void)
{
  signed char cs0   = f();
  signed char cs1   = e();  
  printf("255: %d, EOF: %d\n", (int) cs0, (int)cs1);
  return 0;
}
[yamato@slave]/tmp% gcc foo.c
gcc foo.c
[yamato@slave]/tmp% ./a.out 
./a.out 
255: -1, EOF: -1
```
The parser cannot distinguish 255 and EOF.

If the platform that we use uses unsigned char as "char", we cannot receive value 255. 
```
[yamato@slave]/tmp% cat bar.c
cat bar.c
#include <stdio.h>

int f()
{
  return 255;
}

int e()
{
  return EOF;
}

int main(void)
{
  unsigned char cs0   = f();
  unsigned char cs1   = e();  
  printf("255: %d, EOF: %d\n", (int)cs0, (int)cs1);
  return 0;
}
[yamato@slave]/tmp% gcc bar.c
gcc bar.c
[yamato@slave]/tmp% ./a.out 
./a.out 
255: 255, EOF: 255
```

Either case, int is needed to receive the value returned from the input functions.
See https://github.com/universal-ctags/ctags/pull/1697

**Comment by masatake** [comment]
I read tryParser.

As I wrote, please, change  the type for `c` to `int`?
This request is nothing to do with ungetc issue.

About ungetc,
I will add following thin layer to the main part (you called `core`).

```
marker *pushMarkerToInputStream ();
void       rewindToMarker(marker *);
void      dropMarker(marker *);
```

In tryparser(), you can use the functoins like:

``
int c;
...
marker *m = pushMarkerToInputStream ();
while (result.status == PARSER_NEEDS_MORE_INPUT)
	{
		c = getcFromInputFile ();
                parser (c, token, currentState, &result);
          }
...
if (result.status == PARSER_FAILED)
  rewindToMarker (m);
else
{
   dropMarker(m);
    if (result.unusedChars > 0)
      ungetcToInpuTfile (c);  /* may be one byte is enough
}
...
```
 I have not inspected "else" case well yet.

**Comment by masatake** [comment]
You have to update win32/{ctags_vs2013.vcxproj and ctags_vs2013.vcxproj.filters}, too.

**Comment by ksamborski** [comment]
Thanks! I'll fix it.

**Comment by ksamborski** [comment]
I've changed char to int. The rewinding API you proposed is great. Let me know if I can help you somehow with it.

**Comment by masatake** [comment]
Implementing unwind API is harder than I expected.
I will try to convert your code to token stream based.
I will do it incrementally. #2051 is one of efforts.
Also I wonder whether I can shared the code between TypeScript parser and JavaScript parser.

I will report the progress of conversion.

**Comment by ksamborski** [comment]
> Implementing unwind API is harder than I expected.
> I will try to convert your code to token stream based.
> I will do it incrementally. #2051 is one of efforts.
> Also I wonder whether I can shared the code between TypeScript parser and JavaScript parser.
> 
> I will report the progress of conversion.

Maybe another option would be to change `ungetc` to something like `cppUngetc` is doing but additionally with updating line position?

**Comment by masatake** [comment]
I'm thinking about performance (parsing speed).
Byte stream level backtracking may take much cost, I guess.

**Comment by masatake** [comment]
If all my approaches are failed, I will do as you wrote.

**Comment by ksamborski** [comment]
It depends of the parser. I've tried to design parsers to fail quickly so probably there won't be that much backtracking. But it's a guess indeed.

So I think that with backtracking we get less ifs and clearer code base but parsers may overuse it and then performance will drop. I'm not entirely sure if performance will be much better without it (those ifs and switches have also a cost) but fixing bugs will be much harder for sure.

**Comment by masatake** [comment]
Report of my work.

Though I wanted to convert this one to token stream based, I cannot find time to do so now.
On the other hand, many people wants typescript parser.

So I implemented very thin non-generic unlimitedly ungetc-able input layer in typescript.c.
I would like to call the original version submitted to this pull request `ctags-cppstream`.
I would like to call the one the I replaced the cpp related function with my new thin layer `ctags-tstream`.

I built both binaries and run performance test.
The target is the source code of tslint and typescript language processor.
lcopy files are:
```
[yamato@slave]~/var/codebase% cat lcopy.d/tslint.lcopy 
REPO=https://github.com/palantir/tslint
ALIGNMENT=5.14.0
LANGUAGES=TypeScript
[yamato@slave]~/var/codebase% cat lcopy.d/typescript.lcopy 
REPO=https://github.com/Microsoft/TypeScript
ALIGNMENT=v3.4-rc
LANGUAGES=TypeScript
```
About lcopy files, see universal-ctags/codebase repo.

The following patch is needed to the original one for running ctags-cppstream under codebase:
```
commit 624c3d35e9fdc67a6e24a8cf7b933303769bdd4e (HEAD -> ksamborski-typescript)
Author: Masatake YAMATO <yamato@redhat.com>
Date:   Wed Mar 27 05:11:21 2019 +0900

    TypeScript: init CPreProcessor parser

diff --git a/parsers/typescript.c b/parsers/typescript.c
index 439fc259..40ba47df 100644
--- a/parsers/typescript.c
+++ b/parsers/typescript.c
@@ -61,6 +61,14 @@
 
 extern int cppGetcFromUngetBufferOrFile (void);
 extern void cppUngetc (const int c);
+extern void cppInit (const bool state,
+                    const bool hasAtLiteralStrings,
+                    const bool hasCxxRawLiteralStrings,
+                    const bool hasSingleQuoteLiteralNumbers,
+                        int defineMacroKindIndex,
+                    int macroUndefRoleIndex,
+                    int headerKindIndex,
+                    int headerSystemRoleIndex, int headerLocalRoleIndex);
 
 /*
  *     DATA DEFINITIONS
@@ -1963,6 +1971,9 @@ static void findTsTags (void)
 {
        tokenInfo *const token = newToken ();
 
+       cppInit (false, false, false, false,
+                        KIND_GHOST_INDEX, ROLE_INDEX_DEFINITION,
+                        KIND_GHOST_INDEX, ROLE_INDEX_DEFINITION, ROLE_INDEX_DEFINITION);
        parseTsFile (token);
 
        deleteToken (token);

```

The change for ctags-tsstream is submitted as a pull request #2064.

Parsing speed:
```
[yamato@slave]~/var/codebase% CTAGS_EXE=/home/yamato/var/ctags-typescript/ctags-tsstream ./codebase ctags TypeScript 
version: 2afc9f40
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell +packcc
log: results/2afc9f40,TypeScript..........,..........,time......,default...,2019-03-27-05:21:02.log
tagsoutput: /dev/null
cmdline: + /home/yamato/var/ctags-typescript/ctags-tsstream --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=TypeScript -o - -R code/tslint code/typescript
14553 files, 1282547 lines (28846 kB) scanned in 7.5 seconds (3867 kB/s)
126133 tags added to tag file

real	0m7.481s
user	0m7.298s
sys	0m0.165s
+ set +x
[yamato@slave]~/var/codebase% CTAGS_EXE=/home/yamato/var/ctags-typescript/ctags-cppstream ./codebase ctags TypeScript
version: 624c3d35
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell +packcc
log: results/624c3d35,TypeScript..........,..........,time......,default...,2019-03-27-05:21:23.log
tagsoutput: /dev/null
cmdline: + /home/yamato/var/ctags-typescript/ctags-cppstream --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=TypeScript -o - -R code/tslint code/typescript
14553 files, 1282547 lines (28846 kB) scanned in 6.1 seconds (4714 kB/s)
126131 tags added to tag file

real	0m6.137s
user	0m5.948s
sys	0m0.174s
+ set +x
[yamato@slave]~/var/codebase% CTAGS_EXE=/home/yamato/var/ctags-typescript/ctags-cppstream ./codebase ctags Go        
version: 624c3d35
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell +packcc
log: results/624c3d35,Go..................,..........,time......,default...,2019-03-27-05:21:57.log
tagsoutput: /dev/null
cmdline: + /home/yamato/var/ctags-typescript/ctags-cppstream --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=Go -o - -R code/buildah code/kubernetes
16397 files, 5116001 lines (162732 kB) scanned in 3.4 seconds (48001 kB/s)
627703 tags added to tag file

real	0m3.410s
user	0m3.109s
sys	0m0.290s
+ set +x
[yamato@slave]~/var/codebase% CTAGS_EXE=/home/yamato/var/ctags-typescript/ctags-cppstream ./codebase ctags C 
version: 624c3d35
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell +packcc
log: results/624c3d35,C...................,..........,time......,default...,2019-03-27-05:22:11.log
tagsoutput: /dev/null
cmdline: + /home/yamato/var/ctags-typescript/ctags-cppstream --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=C -o - -R code/linux code/php-src code/ruby
27886 files, 19784832 lines (555082 kB) scanned in 16.0 seconds (34690 kB/s)
1158391 tags added to tag file

real	0m16.069s
user	0m15.540s
sys	0m0.465s
+ set +x
```

ctags-tsstream is 10% slower than ctags-cppstram.
C and Go are popular token based parser.
C parser is 10 times faster than (typescript paser of) ctags-tssream.
Go parser is 10 times faster than (typescript paser of) ctags-cppstream.

Parsing quality:
```
[yamato@slave]~/var/ctags-typescript% cat -n /tmp/foo.ts                           
     1	import * as utils from "tsutils";
     2	import * as ts from "typescript";
     3	
     4	import * as Lint from "../index";
     5	
     6	const OPTION_CHECK_MULTILINE_START = "check-multiline-start";
     7	
     8	export class Rule extends Lint.Rules.AbstractRule {
     9	    /* tslint:disable:object-literal-sort-keys */
    10	    public static metadata: Lint.IRuleMetadata = {
    11	        ruleName: "jsdoc-format",
    12	        description: "Enforces basic format rules for JSDoc comments.",
    13	        descriptionDetails: Lint.Utils.dedent`
    14	            The following rules are enforced for JSDoc comments (comments starting with \`/**\`):
    15	
    16	            * each line contains an asterisk and asterisks must be aligned
    17	            * each asterisk must be followed by either a space or a newline (except for the first and the last)
    18	            * the only characters before the asterisk on each line must be whitespace characters
    19	            * one line comments must start with \`/** \` and end with \`*/\`
    20	            * multiline comments don't allow text after \`/** \` in the first line (with option \`"${OPTION_CHECK_MULTILINE_START}"\`)
    21	        `,
    22	        rationale: "Helps maintain a consistent, readable style for JSDoc comments.",
    23	        optionsDescription: Lint.Utils.dedent`
    24	            You can optionally specify the option \`"${OPTION_CHECK_MULTILINE_START}"\` to enforce the first line of a
    25	            multiline JSDoc comment to be empty.
    26	        `,
    27	        options: {
    28	            type: "array",
    29	            minItems: 0,
    30	            maxItems: 1,
    31	            items: {
    32	                type: "string",
    33	                enum: [OPTION_CHECK_MULTILINE_START],
    34	            },
    35	        },
    36	        optionExamples: [true, [true, OPTION_CHECK_MULTILINE_START]],
    37	        type: "formatting",
    38	        typescriptOnly: false,
    39	    };
    40	    /* tslint:enable:object-literal-sort-keys */
    41	
    42	    public static ALIGNMENT_FAILURE_STRING = "asterisks in jsdoc must be aligned";
    43	    public static FORMAT_FAILURE_STRING = "jsdoc is not formatted correctly on this line";
    44	
    45	    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
    46	        return this.applyWithFunction(sourceFile, walk, {
    47	            firstLineOfMultiline: this.ruleArguments.indexOf(OPTION_CHECK_MULTILINE_START) !== -1,
    48	        });
    49	    }
    50	}
    51	
    52	interface Options {
    53	    firstLineOfMultiline: boolean;
    54	}
[yamato@slave]~/var/ctags-typescript% ./ctags-tsstream --sort=no --fields=+n -o - /tmp/foo.ts
OPTION_CHECK_MULTILINE_START	/tmp/foo.ts	/^const OPTION_CHECK_MULTILINE_START = "check-multiline-start";$/;"	v	line:6
Rule	/tmp/foo.ts	/^export class Rule extends Lint.Rules.AbstractRule {$/;"	c	line:8
Options	/tmp/foo.ts	/^interface Options {$/;"	i	line:52
firstLineOfMultiline	/tmp/foo.ts	/^    firstLineOfMultiline: boolean;$/;"	p	line:53	interface:Options
[yamato@slave]~/var/ctags-typescript% ./ctags-cppstream --sort=no --fields=+n -o - /tmp/foo.ts
OPTION_CHECK_MULTILINE_START	/tmp/foo.ts	/^const OPTION_CHECK_MULTILINE_START = "check-multiline-start";$/;"	v	line:6
Rule	/tmp/foo.ts	/^export class Rule extends Lint.Rules.AbstractRule {$/;"	c	line:8
Options	/tmp/foo.ts	/^}$/;"	i	line:54
firstLineOfMultiline	/tmp/foo.ts	/^}$/;"	p	line:54	interface:Options
```

About pattern fields and line fields for `Options` and `firstLineOfMultilie`, ctags-cppstream emits unusable tags.

Conclusion:
byte-stream based backtracking parser is not so fast as token-based parser.
HOWEVER, as far as KINDs are well designed, it is acceptable; with keeping the design of kinds, we can optimize the implementation or rewrite the parser with the same kinds design.

CPreProcessor layer is not suitable for backtracking. The line numbers for tags take broken value.

**Comment by masatake** [comment]
@ksamborski, do you want to improve more on your side?
For awhile (a month or more), I'm busy so I cannot do much things for ctags.
So you can improve any place in anyway:-)

Anyway, I would like yout to see and try #2064.

**Comment by ksamborski** [comment]
I would like to merge the ctags-tsstream version. It's really great what you have done there. I used the CPreProcessor just as a proof of concept not as a target solution. I think the best case is to merge tsstream version and hopefully I will find some time to rewrite it as incremental parser.

One thing that is better done in cpp version of getc and ungetc is memory allocation. It doesn't allocate new memory for every new character but instead it allocates it for 8 characters at ones. Maybe we could add the same trick here or maybe it's just better to rewrite it as token based parser.

On the other hand there is plenty of work to fully support typescript. But we need to start from somewhere. So if you don't mind, let's just merge the parser (with your changes) and improve it in separate PRs.

**Comment by ksamborski** [comment]
I've updated my branch with your changes. Could we merge it now?

File: Units/parser-typescript.r/ts-namespaces.d/expected.tags [hunk 2]
```diff
@@ -0,0 +1,13 @@
+LettersOnlyValidator	input.ts	/^    export class LettersOnlyValidator implements StringValidator {$/;"	c
+StringValidator	input.ts	/^    export interface StringValidator {$/;"	i	namespace:Validation
+Validation	input.ts	/^namespace Validation {$/;"	n
+ZipCodeValidator	input.ts	/^    export class ZipCodeValidator implements StringValidator {$/;"	c
+isAcceptable	input.ts	/^        isAcceptable(s: string) {$/;"	m	class:LettersOnlyValidator
+isAcceptable	input.ts	/^        isAcceptable(s: string) {$/;"	m	class:ZipCodeValidator
+isAcceptable	input.ts	/^        isAcceptable(s: string): boolean;$/;"	m	interface:Validation.StringValidator
+lettersRegexp	input.ts	/^    const lettersRegexp = \/^[A-Za-z]+$\/;$/;"	v
+numberRegexp	input.ts	/^    const numberRegexp = \/^[0-9]+$\/;$/;"	v
```
[the remark this answers is not in the document]
Comment (ksamborski) [reply Units/parser-typescript.r/ts-namespaces.d/expected.tags]: It depends how you understand it. In typescript (and in JS) it generally means that it cannot be reassigned but if it's an array or object then its elements can be changed. So it's not that constant in my opinion. And I don't know if it was a good decision by my side but it was just easier to handle it like any other variable. What do you think?

**Comment by masatake** [comment]
I guess test cases are taken from https://www.typescriptlang.org/docs/handbook/.
Am I correct?

File: Units/parser-typescript.r/ts-namespaces.d/expected.tags (hunk 2 above)
[the remark this answers is not in the document]
Comment (masatake) [reply Units/parser-typescript.r/ts-namespaces.d/expected.tags]: Introducing `const` is better.
Converting a const kind object to a variable kind object can be done in a client tool at an upper layer. "Just provide low information to clients" motto of u-ctags.

**Comment by ksamborski** [comment]
> I guess test cases are taken from https://www.typescriptlang.org/docs/handbook/.
> Am I correct?
> 

Only some of them. The rest I added when I encountered problems with tags when working with typescript at work.

File: Units/parser-typescript.r/ts-namespaces.d/expected.tags (hunk 2 above)
[the remark this answers is not in the document]
Comment (ksamborski) [reply Units/parser-typescript.r/ts-namespaces.d/expected.tags]: OK, I'll change it.

[the remark this answers is not in the document]
Comment (ksamborski) [reply Units/parser-typescript.r/ts-namespaces.d/expected.tags]: I've added support for constants using `C` kind.

**Comment by masatake** [comment]
>> I guess test cases are taken from https://www.typescriptlang.org/docs/handbook/.
Am I correct?

> Only some of them. The rest are based on my code and I added when I encountered problems with tags when working with typescript at work.

Thank you for responding. I will add README at Units/parser-typescript with following contents:
```
About input files, some of them are taken from https://www.typescriptlang.org/docs/handbook/
```

If I did something what you don't intend, please let me know.

File: Units/parser-typescript.r/ts-namespaces.d/expected.tags (hunk 2 above)
[the remark this answers is not in the document]
Comment (masatake) [reply Units/parser-typescript.r/ts-namespaces.d/expected.tags]: Thank you. I will pick up the change to #2064.
@felixfbecker, thank you. This kind of comments is really welcome.

**Comment by ksamborski** [comment]
> ```
> About input files, some of them are taken from https://www.typescriptlang.org/docs/handbook/
> ```

Not whole input files but only some parts of them are taken from official docs. I think most sample code parts are taken from: https://github.com/Microsoft/TypeScript/blob/master/doc/TypeScript%20Language%20Specification.pdf

**Comment by masatake** [comment]
How about following sentence?

```
Not whole input files but only some parts of them are taken from official docs[1]
[1] https://github.com/Microsoft/TypeScript/blob/master/doc/TypeScript%20Language%20Specification.pdf
```

**Comment by ksamborski** [comment]
Fine by me :)

**Comment by masatake** [comment]
Merged via #2064.

**Comment by masatake** [comment]
Your parser eliminates typescript.ctags from sourcegraph!
https://github.com/sourcegraph/sourcegraph/pull/4987/commits/b62f49274d51121c5e9b02005b03677a45e18a8c

**Pull request closed without merging**